minetest/minetest
# WIP: Performance improvements for entity collision

@appgurueu suggested that entity collision detection can be improved using a spatial index. This is our combined effort to make it happen.
 
Fixes #6453

Pros:
    - Servers will be able to handle more entities without getting as bogged down. One example usecase (credit due to Mister E) is advtrains.
    - We are not going to be changing behavior of the existing Lua API. We are only speeding up the implementation.
    - The Lua API can be easily extended to support new types of collision
    - The whole spatial index code is getting a nice refactor and will be properly tested.

Cons:
    ~~- We're planning to add spatial index as a required dependency; we know this isn't a light decision. The alternative is to rely on a vector based replacement, which forgoes the performance gains for significant extra work in code maintenance.~~

THST has been chosen for the spatial indexing library because it's header only and can very easily be extended to do line intersection. It has somewhat better documentation than libspatialindex, and has unit tests. libspatialindex will be removed as a dependency altogether.

@appgurueu may have more things to add that I missed.

## To do

This PR is a Work in Progress

- [x] Add spatial store template with unit testing
- [ ] Add unit test for and fix line segment collision
- [x] Fix CI builds (except Android for now) and comply with clang format and clang tidy
- [ ] Add a PointStore
- [ ] Refactor AreaStore to use the spatial store
- [ ] Update all entity collision to use the spatial store
- [ ] Convince sfan5 to let me use uniform initialization

## How to test

Use along with #12040

The spatial store is unit tested. In-game, you can observe that collision works as expected. We have a mod for testing the performance. I will link it here once a "tryable" state is reached.
